pydantic/_internal/_utils.py

```diff
@@ -333,3 +335,43 @@ def all_identical(left: typing.Iterable[Any], right: typing.Iterable[Any]) -> bo
         if left_item is not right_item:
             return False
     return True
+
+
+_KeyType = TypeVar('_KeyType')
+_ValueType = TypeVar('_ValueType')
+
+
+# We need a sentinel value for missing keys that cannot be None, since None is a legitimate value
+# This sentinel must be hashable, compare equal only to itself, and properly typeable in type annotations
+# We use the single-value enum trick to allow correct typing when using a sentinel, from the MyPy suggestion:
+# https://github.com/python/typing/issues/689#issuecomment-561568944
+class SentinelType(enum.Enum):
+    """Class for a sentinel value so that sentinel usage can written and type-checked in type annotations"""
+
+    SENTINEL = enum.auto()
+
+
+SENTINEL = SentinelType.SENTINEL
+"""Sentinel value that can be properly typed in type annotations"""
+
+
+@dataclasses.dataclass(frozen=True)
+class SafeGetItemProxy(Generic[_KeyType, _ValueType]):
+    """Wrapper redirecting `__getitem__` to `get` with a sentinel value as default
+
+    This makes is safe to use in `operator.itemgetter` when some keys may be missing
+    """
+
+    wrapped: Mapping[_KeyType, _ValueType]
+
+    def __getitem__(self, __key: _KeyType) -> SentinelType | _ValueType:
+        return self.wrapped.get(__key, SENTINEL)
+
+    # required to pass the object to operator.itemgetter() instances due to a quirk of typeshed
+    # https://github.com/python/mypy/issues/13713
+    # https://github.com/python/typeshed/pull/8785
+    # Since this is typing-only, hide it in a typing.TYPE_CHECKING block
+    if typing.TYPE_CHECKING:
+
+        def __contains__(self, __key: _KeyType) -> bool:
+            return self.wrapped.__contains__(__key)
```

Comment:
I would suggest changing this to something like:
```python
import dataclasses
import typing
from typing import Any

from ._internal import _decorators, _internal_dataclass


_SENTINEL = object()

@dataclasses.dataclass(**_internal_dataclass.slots_true, frozen=True)
class SafeGetItemProxy:
    """Wrapper redirecting `__getitem__` to `get` with a sentinel value as default
    This makes is safe to use in `operator.itemgetter` when some keys may be missing
    """

    wrapped: dict[str, Any]

    def __getitem__(self, __key: str) -> Any:
        return self.wrapped.get(__key, _SENTINEL)

    # required to pass the object to operator.itemgetter() instances due to a quirk of typeshed
    # https://github.com/python/mypy/issues/13713
    # https://github.com/python/typeshed/pull/8785
    # Since this is typing-only, hide it in a typing.TYPE_CHECKING block
    if typing.TYPE_CHECKING:

        def __contains__(self, __key: str) -> bool:
            return self.wrapped.__contains__(__key)
```
Obviously you won't need all those imports etc.

Reply:
That makes perfect sense, good point that this is only used at place where types are already Any. I'll update.

Should I merge the `_EMPTY` and `_SENTINEL` that are in that file? Looks like I can just reuse `_EMPTY`, perhaps renaming it?

Reply:
What do you think about defining `__slots__` manually instead of using the special switch? IIRC it would work for all versions of python and provide the slightly faster attribute access?